HaloMD/master_server/src/heartbeat_packet.rs

```diff
@@ -0,0 +1,77 @@
+// This isn't the whole packet, just the stuff we care about.
+pub struct HeartbeatPacket {
+    pub localport: u16,     // Can't trust source port due to some routers using a different port than this.
+    pub gamename: String,   // Must be "halor"
+    pub gamever: String,    // Must be 1.09 or 1.10 in case it's unjoinable.
+    pub statechanged: u16   // If it's 2, then the server is shutting down.
+}
+
+// C strings are null terminated, so they end with a 0 byte.
+fn get_string_from_c_string(buffer: &[u8]) -> String {
+    let mut bytes : Vec<u8> = Vec::new();
+    for i in buffer {
+        if *i == 0 {
+            break;
+        }
+        bytes.push(*i);
+    }
+    return match String::from_utf8(bytes) {
+        Err(_) => String::new(),
+        Ok(n) => n
+    }
+}
+
+impl HeartbeatPacket {
+    pub fn from_buffer(buffer: &[u8]) -> HeartbeatPacket {
+        let mut ret = HeartbeatPacket {localport: 0, gamename: "".to_string(), gamever: "".to_string(), statechanged: 0};
+
+        // Heartbeat packets are more than 5 bytes. The buffer should be bigger than the actual length we need.
+        if buffer.len() < 5 || buffer[0] != 3 {
+            return ret;
+        }
+
+        let mut offset = 5;
+        let mut last_string = String::new();
+
+        loop {
+            if offset >= buffer.len() {
+                break;
+            }
+
+            let value = get_string_from_c_string(&buffer[offset..buffer.len()]);
+
```

Comment:
Try to also grab the game key and game value in the same loop block/iteration. This would involve the iterator being advanced twice (`next()`); not sure if the fancy for loop would work for this.